File: modules/secure-recorder/ios/Session.swift
import Foundation
import AVFoundation

/**
 * Manages a single recording session for iOS/iPadOS
 * 
 * Orchestrates recording components (Facade pattern). Coordinates audio recording,
 * encryption, timing, and event handling. Manages session lifecycle from
 * initialization through cleanup.
 * 
 * iOS/iPadOS SPECIFICITY:
 * - Uses Swift async/await for asynchronous operations
 * - start() throws SecureRecorderError
 * - Uses AVAudioEngine/AVAudioInputNode (via AudioRecorder)
 * - Separate initializeEventHandler() method (needed for weak self capture)
 * - getInfo() returns tuple (sessionId, filePath, isActive)
 * - Explicitly removes existing file before recording (FileManager.removeItem)
 * - cleanup() has no try-catch (simpler cleanup, errors are non-critical)
 */
class Session {
  private let keyManager: KeyManagerProtocol
  private let audioRecorder: AudioRecorder
  private let audioConfig: AudioConfig
  private let limiter: LimitRegistryProtocol
  private let onLimitReached: ((StopReason, String, String) async -> Void)?
  private let sessionId: String
  private let outputFile: URL
  private var encryptionStream: EncryptionStreamProtocol!
  private var audioRecord: AudioRecord!
  private var eventHandler: EventHandler!
  private var pipeline: Pipeline?
  private var recordingQueue: DispatchQueue?
  
  internal let recordingTimer = RecordingTimer()
  
  internal init(
    sessionId: String,
    outputFile: URL,
    keyManager: KeyManager,
    audioRecorder: AudioRecorder,
    audioConfig: AudioConfig,
    limiter: LimitRegistry = LimitRegistry(),
    onLimitReached: ((StopReason, String, String) async -> Void)? = nil
  ) {
    self.sessionId = sessionId
    self.outputFile = outputFile
    self.keyManager = keyManager
    self.audioRecorder = audioRecorder
    self.audioConfig = audioConfig
    self.limiter = limiter
    self.onLimitReached = onLimitReached
  }
  
  private func initializeEventHandler() {
    eventHandler = EventHandler(
      sessionId: sessionId,
      outputFile: outputFile,
      recordingTimer: recordingTimer,
      onStop: { [weak self] in
        guard let self = self else { throw SecureRecorderError.initializationFailed("Session deallocated") }
        return try self.stop()
      },
      onLimitReached: onLimitReached
    )
  }
  
  /**
   * Start recording session
   * Initializes encryption and audio capture
   * 
   * @return File path where encrypted audio is being written
   * @throws SecureRecorderError if initialization fails
   */
  internal func start(keyAlias: String) throws -> String {
    let startTs = Date()
    print("[SecureRecorder][Session] start requested, sessionId=\(sessionId) at \(startTs)")
    // Get or create encryption key
    let key = try keyManager.getOrCreateKey(alias: keyAlias)
    
    // Remove file ONLY on fresh start (not resume)
    if FileManager.default.fileExists(atPath: outputFile.path) {
      try FileManager.default.removeItem(at: outputFile)
    }
    
    // Initialize encryption stream
    encryptionStream = EncryptionStream(secretKey: key, outputFile: outputFile)
    try encryptionStream.initialize()
    
    try startAudioCapture()
    
    return outputFile.path
  }
  
  /**
   * Start audio capture and pipeline processing
   * 
   * Extracted from start() to allow reuse by resume().
   * Creates pipeline, starts audio recording, and launches processing loop.
   */
  private func startAudioCapture() throws {
    audioRecord = try audioRecorder.start()
    
    initializeEventHandler()
    
    recordingTimer.activate()
    
    pipeline = Pipeline(
      audioRecord: audioRecord,
      encryptionStream: encryptionStream,
      audioConfig: audioConfig,
      outputFile: outputFile,
      limiter: limiter,
      recordingTimer: recordingTimer,
      onLimitReached: { [weak self] reason in
        self?.eventHandler.onLimitReached(reason: reason)
      },
      onError: { [weak self] message in
        self?.eventHandler.onError(message: message)
      }
    )
    
    try audioRecord.startRecording()
    
    guard let pipelineRef = pipeline else {
      recordingTimer.deactivate()
      throw SecureRecorderError.initializationFailed("Pipeline not initialized")
    }
    
    let queue = DispatchQueue(label: "com.tiroscribe.secure-recorder.pipeline", qos: .userInitiated)
    recordingQueue = queue
    
    queue.async { [weak pipelineRef] in
      pipelineRef?.process()
    }
  }
  
  /**
   * Pause recording session
   * Stops audio capture but keeps encryption stream open for resume
   * 
   * @return File path where encrypted audio is written
   * @throws SecureRecorderError if pause fails
   */
  internal func pause() throws -> String {
    guard recordingTimer.isActive else {
      throw SecureRecorderError.noRecordingInProgress
    }
    
    // Stop audio recording
    if audioRecord != nil {
      audioRecorder.stop(record: audioRecord)
    }
    
    // Flush buffered audio to disk (encryption stream stays open)
    try encryptionStream.flush()
    
    recordingTimer.deactivate()
    pipeline = nil
    recordingQueue = nil
    audioRecord = nil
    
    return outputFile.path
  }
  
  /**
   * Resume recording session
   * Restarts audio capture to existing encryption stream
   * 
   * @return File path where encrypted audio is written
   * @throws SecureRecorderError if resume fails
   */
  internal func resume() throws -> String {
    guard !recordingTimer.isActive else {
      throw SecureRecorderError.recordingInProgress
    }
    
    guard encryptionStream != nil else {
      throw SecureRecorderError.initializationFailed("No encryption stream to resume")
    }
    
    try startAudioCapture()
    
    return outputFile.path
  }
  
  /**
   * Stop recording session
   * Finalizes encryption and stops audio capture
   * 
   * @return File path where encrypted audio was written
   * @throws SecureRecorderError if finalization fails
   */
  internal func stop() throws -> String {
    let stopTs = Date()
    print("[SecureRecorder][Session] stop requested, sessionId=\(sessionId) at \(stopTs)")
    // Stop audio recording (isomorphic: matches Android AudioRecorder.stop())
    if audioRecord != nil {
      audioRecorder.stop(record: audioRecord)
    }
    
    // Close encryption stream safely
    if encryptionStream != nil {
      encryptionStream.close()
    }
    
    // Deactivate timer
    recordingTimer.deactivate()
    
    audioRecord = nil
    pipeline = nil
    recordingQueue = nil
    
    return outputFile.path
  }
  
  /**
   * Cleanup session resources
   * Safe to call even if session wasn't fully initialized
   */
  internal func cleanup() {
    if audioRecord != nil {
      audioRecorder.stop(record: audioRecord)
    }
    
    // Close encryption stream safely
    if encryptionStream != nil {
      encryptionStream.close()
    }
    
    recordingTimer.deactivate()
    audioRecord = nil
    pipeline = nil
    recordingQueue = nil
  }
  
  /**
   * Get session information
   */
  internal func getInfo() -> (sessionId: String, filePath: String, isActive: Bool) {
    return (sessionId, outputFile.path, recordingTimer.isActive)
  }
}

File: modules/secure-recorder/android/src/main/java/expo/modules/securerecorder/Session.kt
package expo.modules.securerecorder

import android.media.AudioRecord
import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.launch
import java.io.File

/**
 * Manages a single recording session for Android
 * 
 * Orchestrates recording components (Facade pattern). Coordinates audio recording,
 * encryption, state management, and event handling. Manages session lifecycle from
 * initialization through cleanup.
 * 
 * ISOMORPHIC: Matches iOS Session exactly
 * - Both: start, pause, resume, stop, cleanup, startAudioCapture
 * - Both: 1 encounter = 1 file (pause keeps file open)
 * 
 * ANDROID SPECIFICITY:
 * - Uses Kotlin coroutines (CoroutineScope) for asynchronous operations
 * - start() is synchronous (no throws)
 * - Uses AudioRecord API (via AudioRecorder)
 * - EventHandler initialized inline in start()
 * - getInfo() returns SessionInfo data class
 * - File overwrites automatically (FileOutputStream)
 * - cleanup() has try-catch for best-effort cleanup
 */
class Session(
  private val sessionId: String,
  private val outputFile: File,
  private val keyManager: KeyManagerInterface,
  private val audioRecorder: AudioRecorder,
  private val audioConfig: AudioConfig,
  private val limiter: LimitRegistryInterface = LimitRegistry(),
  private val onLimitReached: (suspend (StopReason, String, String) -> Unit)? = null
) {
  private lateinit var eventHandler: EventHandler
  private lateinit var encryptionStream: EncryptionStream
  private lateinit var currentAudioRecord: AudioRecord
  private var recordingJob: Job? = null
  private val recordingScope = CoroutineScope(SupervisorJob() + Dispatchers.IO)
  
  internal val recordingTimer = RecordingTimer()
  
  /**
   * Start recording session
   * Initializes encryption and audio capture
   * 
   * @return File path where encrypted audio is being written
   * @throws Exception if initialization fails
   */
  internal fun start(keyAlias: String): String {
    val startTs = System.currentTimeMillis()
    Log.d("SecureRecorder", "Session.start invoked sessionId=$sessionId at $startTs")
    // Initialize event handler (needs stop() method, so initialize here)
    eventHandler = EventHandler(
      sessionId = sessionId,
      outputFile = outputFile,
      recordingTimer = recordingTimer,
      onStop = { stop() },
      onLimitReached = onLimitReached
    )
    
    // Get or create encryption key
    val secretKey = keyManager.getOrCreateKey(keyAlias)
    
    // Remove file ONLY on fresh start (not resume)
    if (outputFile.exists()) {
      outputFile.delete()
    }
    
    // Initialize encryption stream
    encryptionStream = EncryptionStream(secretKey, outputFile)
    encryptionStream.initialize()
    
    startAudioCapture()
    
    return outputFile.absolutePath
  }
  
  /**
   * Start audio capture and pipeline processing
   * 
   * Extracted from start() to allow reuse by resume().
   * Creates pipeline, starts audio recording, and launches processing coroutine.
   */
  private fun startAudioCapture() {
    // Start audio recording
    val audioRecord = audioRecorder.start()
    audioRecord.startRecording()
    currentAudioRecord = audioRecord
    
    // Activate state
    recordingTimer.activate()
    
    // Create pipeline for audio processing
    val pipeline = Pipeline(
      audioRecord = audioRecord,
      encryptionStream = encryptionStream,
      audioConfig = audioConfig,
      outputFile = outputFile,
      limiter = limiter,
      recordingTimer = recordingTimer,
      onLimitReached = { reason ->
        eventHandler.onLimitReached(reason)
      },
      onError = { message ->
        eventHandler.onError(message)
      }
    )
    
    // Start recording coroutine
    recordingJob = recordingScope.launch {
      pipeline.process()
    }
  }
  
  /**
   * Pause recording session
   * Stops audio capture but keeps encryption stream open for resume
   * 
   * @return File path where encrypted audio is written
   * @throws NoRecordingException if not currently recording
   */
  internal fun pause(): String {
    if (!recordingTimer.isActive) {
      throw expo.modules.securerecorder.exception.NoRecordingException()
    }
    
    // Cancel recording job
    recordingJob?.cancel()
    recordingJob = null
    
    // Stop audio recording
    if (::currentAudioRecord.isInitialized) {
      audioRecorder.stop(currentAudioRecord)
    }
    
    // Flush buffered audio to disk (encryption stream stays open)
    encryptionStream.flush()
    
    // Deactivate state
    recordingTimer.deactivate()
    
    return outputFile.absolutePath
  }
  
  /**
   * Resume recording session
   * Restarts audio capture to existing encryption stream
   * 
   * @return File path where encrypted audio is written
   * @throws RecordingInProgressException if already recording
   * @throws InitializationException if no encryption stream to resume
   */
  internal fun resume(): String {
    if (recordingTimer.isActive) {
      throw expo.modules.securerecorder.exception.RecordingInProgressException()
    }
    
    if (!::encryptionStream.isInitialized) {
      throw expo.modules.securerecorder.exception.InitializationException("No encryption stream to resume")
    }
    
    startAudioCapture()
    
    return outputFile.absolutePath
  }
  
  /**
   * Stop recording session
   * Finalizes encryption and stops audio capture
   * 
   * @return File path where encrypted audio was written
   * @throws Exception if finalization fails
   */
  internal fun stop(): String {
    val stopTs = System.currentTimeMillis()
    Log.d("SecureRecorder", "Session.stop invoked sessionId=$sessionId at $stopTs")
    // Cancel recording job
    val job = recordingJob
    if (job != null) {
      job.cancel()
      recordingJob = null
    }
    
    // Stop audio recording
    if (::currentAudioRecord.isInitialized) {
      audioRecorder.stop(currentAudioRecord)
    }
    
    // Close encryption stream safely
    if (::encryptionStream.isInitialized) {
      encryptionStream.close()
    }
    
    // Deactivate state
    recordingTimer.deactivate()
    
    return outputFile.absolutePath
  }
  
  /**
   * Cleanup session resources
   * Safe to call even if session wasn't fully initialized
   */
  internal fun cleanup() {
    try {
      // Cancel recording job
      val job = recordingJob
      if (job != null) {
        job.cancel()
        recordingJob = null
      }
      
      // Cancel entire scope to prevent resource leaks
      recordingScope.cancel()
      
      // Stop audio recording
      if (::currentAudioRecord.isInitialized) {
        audioRecorder.stop(currentAudioRecord)
      }
      
      // Close encryption stream safely
      if (::encryptionStream.isInitialized) {
        encryptionStream.close()
      }
      
      // Deactivate state
      recordingTimer.deactivate()
    } catch (e: Exception) {
      // Ignore cleanup errors - best effort cleanup
      // State is still updated to prevent further operations
      recordingTimer.deactivate()
    }
  }
  
  /**
   * Get session information
   */
  internal fun getInfo(): SessionInfo {
    return SessionInfo(sessionId, outputFile.absolutePath, recordingTimer.isActive)
  }
  
  data class SessionInfo(
    internal val sessionId: String,
    internal val filePath: String,
    internal val isActive: Boolean
  )
}
